vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c

```diff
@@ -1083,11 +752,294 @@ BTStatus_t prvBTSendResponse( uint16_t usConnId,
+
+                if( pDescriptors->uuid != NULL )
+                {
+                    vPortFree( ( void * ) pDescriptors->uuid );
+                }
+
+                dscrCount++;
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    if( pSvc->characteristics != NULL )
+    {
+        vPortFree( ( void * ) pSvc->characteristics );
+    }
+}
+
+
+/* @brief Simple function that creates a full service in one go.
+ * The function is big because of error checking but in reality it only does the following:
+ * 1. Populate service fields
+ * 2. Populate each attributes by looping into pxService.
+ * An array for service is available in static memory, however, UUIDs and attributes have to be dynamically allocated.
+ * This structure is a compromise between full dynamic allocation and static allocation.
+ */
 BTStatus_t prvAddServiceBlob( uint8_t ucServerIf,
                               BTService_t * pxService )
 {
-    return eBTStatusUnsupported;
+    uint16_t index;
+    uint16_t charCount;
+    uint16_t dscrCount;
+    struct ble_gatt_chr_def * pCharacteristics = NULL;
+    struct ble_gatt_dsc_def * pDescriptors = NULL;
+    BTStatus_t xStatus = eBTStatusSuccess;
+    ble_uuid_t * uuid;
+    struct ble_gatt_svc_def * pSvc = NULL;
+
+    if( serviceCnt < MAX_SERVICES )
+    {
+        pSvc = &espServices[ serviceCnt ];
+        afrServices[ serviceCnt ] = pxService;
+    }
+    else
+    {
+        xStatus = eBTStatusNoMem;
+    }
+
+    /* Fill in service field. After that start creating characteristics and descriptors individually. */
+    if( xStatus == eBTStatusSuccess )
+    {
+        espServices[ serviceCnt + 1 ].type = BLE_GATT_SVC_TYPE_END;
+
+        if( pxService->pxBLEAttributes[ 0 ].xAttributeType == eBTDbPrimaryService )
+        {
+            pSvc->type = BLE_GATT_SVC_TYPE_PRIMARY;
+        }
+        else if( pxService->pxBLEAttributes[ 0 ].xAttributeType == eBTDbPrimaryService )
+        {
+            pSvc->type = BLE_GATT_SVC_TYPE_SECONDARY;
+        }
+        else
+        {
+            configPRINTF( ( "Invalid service type \n" ) );
+            xStatus = eBTStatusFail;
+        }
+    }
+
+    if( xStatus == eBTStatusSuccess )
+    {
+        pSvc->uuid = prvCopytoESPUUID( &pxService->pxBLEAttributes[ 0 ].xServiceUUID, NULL );
+
+        if( pSvc->uuid == NULL )
+        {
+            configPRINTF( ( "Could not allocate memory for service UUID \n" ) );
+            xStatus = eBTStatusNoMem;
+        }
+        else
+        {
+            pCharacteristics = pvPortCalloc( prvCountCharacteristics( pxService ) + 1, sizeof( struct ble_gatt_chr_def ) );
+
+            if( pCharacteristics == NULL )
+            {
+                configPRINTF( ( "Could not allocate memory for  characteristic array \n" ) );
+                prvCleanupService( pxService, pSvc );
+                xStatus = eBTStatusNoMem;
```

Comment:
Maybe I missed something. I did not find where memory allocation is cleaned up, in the case that --
- prvCopytoESPUUID() succeeded. 
-  pCharacteristics = pvPortCalloc() failed.
The second allocation needs to be cleaned up.

Reply:
Good catch too! Fixed